protocols/kad/src/bootstrap.rs

```diff
@@ -0,0 +1,344 @@
+use futures::FutureExt;
+use std::task::{Context, Poll, Waker};
+use std::time::Duration;
+
+use futures_timer::Delay;
+
+#[derive(Debug)]
+pub(crate) struct Status {
+    /// If the user did not disable periodic bootstrap (by providing `None` for `periodic_interval`)
+    /// this is the periodic interval and the delay of the current period. When `Delay` finishes,
+    /// a bootstrap will be triggered and the `Delay` will be reset.
+    interval_and_delay: Option<(Duration, Delay)>,
+
+    /// Configured duration to wait before triggering a bootstrap when a new peer
+    /// is inserted in the routing table. `None` if automatic bootstrap is disabled.
+    automatic_throttle: Option<Duration>,
```

Comment:
I don't really understand why we need this? Can somebody explain please? Why can't we just check if we are currently bootstrapping and not trigger another bootstrap in that case?

Reply:
Even if we need it, I'd suggest to not make this configurable and just run with a default value for now. Or what is the usecase for making this configurable?

Reply:
I think your point is valid and I am happy to add such a throttling but I'd like to do so without a configuration value so we can more easily change this later.

Reply:
Oh ok. To limit potential breaking changes I suppose ? I can understand and I won't mind since it is always easy to make it configurable in a fork if the end user wants to.

Do you have a specific default value in mind ? Maybe `10ms` is a lot... I don't really know.

Reply:
> To limit potential breaking changes I suppose ?

Correct! Anything that is an implementation detail we can fix and release in patch versions, for example alter the value in case whatever we choose doesn't work! :)

> Do you have a specific default value in mind ? Maybe `10ms` is a lot... I don't really know.

Some rudimentary thoughts:

- The RTT of a connection between a client somewhere (mobile, home network, etc) and a server in a datacenter is probably between 10ms and 150ms.
- Not all connections will be established at the same time.

If we set the throttling to 500ms, I think it is reasonable to assume that any "burst" of connections is complete and we can start the bootstrap. At the same time, delaying it by 500ms should be barely noticeable in the context of the application boot.

Reply:
Ok great. I've open a PR against @PanGan21's fork : https://github.com/PanGan21/rust-libp2p/pull/2

I did not completely removed the method `set_automatic_bootstrap_throttle` because it is useful to disable automatic bootstrap in some unit tests (specially in `protocols/kad/src/behaviour/test.rs`